src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#pragma once
+
+#include "bucket/BucketListSnapshot.h"
+#include "bucket/BucketManager.h"
+#include "bucket/BucketSnapshotManager.h"
+#include "ledger/LedgerTxn.h"
+#include "util/NonCopyable.h"
+
+namespace stellar
+{
+
+class Application;
+class TransactionFrame;
+
+// A unified ledger entry interface that supports LedgerEntry representations
+// for both legacy SQL and BucketList snapshots. When working with LedgerTxn,
+// using plain LedgerEntry is not safe, as it might be modified or invalidated
+// by a nested transaction. To address this, the LedgerTxnEntry and
+// ConstLedgerTxnEntry abstractions should be preserved, allowing bucket
+// snapshot LedgerEntries to support the same interface.
+class LedgerEntryWrapper
+{
+    // Either hold a reference or a pointer to the entry
+    std::variant<LedgerTxnEntry, ConstLedgerTxnEntry,
+                 std::shared_ptr<LedgerEntry>>
+        mEntry;
+
+  public:
+    LedgerEntryWrapper(ConstLedgerTxnEntry&& entry);
+    LedgerEntryWrapper(LedgerTxnEntry&& entry);
+    LedgerEntryWrapper(std::shared_ptr<LedgerEntry> entry);
+    LedgerEntry& currentToModify();
+    LedgerEntry const& current() const;
+    operator bool() const;
+};
+
+// A unified ledger header access interface, similar to
+// LedgerEntryWrapper. Just like LedgerEntryWrapper, this class is purely
+// cosmetic for BucketList snapshots, since those are immutable.
+class LedgerHeaderWrapper
+{
+    std::variant<LedgerTxnHeader, std::shared_ptr<LedgerHeader>> mHeader;
+
+  public:
+    LedgerHeaderWrapper(LedgerTxnHeader&& header);
+    LedgerHeaderWrapper(LedgerHeader const& header);
+    LedgerHeaderWrapper(std::shared_ptr<LedgerHeader> header);
+    LedgerHeader& currentToModify();
+    LedgerHeader const& current() const;
+    LedgerTxnHeader const&
+    maybeGetLedgerTxnHeader() const
+    {
+        releaseAssert(std::holds_alternative<LedgerTxnHeader>(mHeader));
+        return std::get<0>(mHeader);
+    }
+};
+
+// A unified interface for read-only ledger state snapshot.
+// Supports SQL (via read-only LedgerTxn), as well as BucketList snapshots.
+class AbstractLedgerStateSnapshot
+{
+  public:
+    virtual ~AbstractLedgerStateSnapshot() = default;
+    virtual LedgerHeaderWrapper getLedgerHeader() = 0;
```

Comment:
Yeah, tbh I don't want `getHeader` to be present in this interface at all, because it's unsafe to expose it like this. The problem is using getLedgerHeader universally doesn't work right now because of https://github.com/stellar/stellar-core/issues/4412, but I didn't get a chance to fix that yet. Maybe I'll push a fix to this PR so we can remove `getHeader` completely.

Reply:
Yeah I 2nd this, if the scope is too large for this PR a comment would be helpful in the meantime, both here in the header and also in the few places we need to call `getHeader`.

Reply:
Agreed. Renamed this function to `getLedgerHeaderUnsafe` to indicate that we shouldn't be using it for now, but I can clean this up in a follow-up.